core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java

```diff
@@ -68,11 +71,38 @@ private static class ConsistentHashProvider {
     private List<BlockWorkerInfo> mLastWorkerInfos = ImmutableList.of();
     private NavigableMap<Integer, BlockWorkerInfo> mActiveNodesByConsistentHashing;
 
+    private volatile long mLastUpdatedTimestamp = 0L;
+
+    private final AtomicBoolean mNeedUpdate = new AtomicBoolean(false);
+
+    private static final long WORKER_INFO_UPDATE_INTERVAL_MS = 30_000L;
+
     public void refresh(List<BlockWorkerInfo> workerInfos, int numVirtualNodes) {
-      if (workerInfos != mLastWorkerInfos
-          && !ImmutableSet.copyOf(workerInfos).equals(ImmutableSet.copyOf(mLastWorkerInfos))) {
-        build(workerInfos, numVirtualNodes);
+      // check if we need to update worker info
+      if (mLastUpdatedTimestamp <= 0L
+          || System.currentTimeMillis() - mLastUpdatedTimestamp > WORKER_INFO_UPDATE_INTERVAL_MS) {
+        mNeedUpdate.set(true);
+      }
+      // update worker info if needed
+      if (mNeedUpdate.compareAndSet(true, false)) {
+        if (isWorkerInfoUpdated(workerInfos, mLastWorkerInfos)) {
+          build(workerInfos, numVirtualNodes);
+        }
+        mLastUpdatedTimestamp = System.currentTimeMillis();
```

Comment:
move this to `build`? so that mutable states are updated in a single place.

Reply:
We can't move it to `build` as build may not be called.